sdk/nodejs/output.ts

```diff
@@ -389,49 +398,28 @@ export function isSecretOutput<T>(o: Output<T>): Promise<boolean> {
     return Output.isInstance(o.isSecret) ? Promise.resolve(false) : o.isSecret;
 }
 
-/**
- * [output] takes any Input value and converts it into an Output, deeply unwrapping nested Input
- * values as necessary.
- *
- * The expected way to use this function is like so:
- *
- * ```ts
- *      var transformed = pulumi.output(someVal).apply(unwrapped => {
- *          // Do whatever you want now.  'unwrapped' will contain no outputs/promises inside
- *          // here, so you can easily do whatever sort of transformation is most convenient.
- *      });
- *
- *      // the result can be passed to another Resource.  The dependency information will be
- *      // properly maintained.
- *      var someResource = new SomeResource(name, { data: transformed ... });
- * ```
- */
-export function output<T>(val: Input<T>): Output<Unwrap<T>>;
-export function output<T>(val: Input<T> | undefined): Output<Unwrap<T | undefined>>;
-export function output<T>(val: Input<T | undefined>): Output<Unwrap<T | undefined>> {
+interface ValueInfo {
+    value: any;
+    isOutput?: boolean;
+}
+
+function outputRec(val: any): ValueInfo {
     if (val === null || typeof val !== "object") {
         // strings, numbers, booleans, functions, symbols, undefineds, nulls are all returned as
         // themselves.  They are always 'known' (i.e. we can safely 'apply' off of them even during
         // preview).
-        return createSimpleOutput(val);
+        return { value: val };
     }
     else if (Resource.isInstance(val)) {
         // Don't unwrap Resources, there are existing codepaths that return Resources through
         // Outputs and we want to preserve them as is when flattening.
-        return createSimpleOutput(val);
+        return { value: val };
     }
     else if (isUnknown(val)) {
-        // Turn unknowns into unknown outputs.
-        return <any>new Output(
-            new Set(), Promise.resolve(<any>val), /*isKnown*/ Promise.resolve(false), /*isSecret*/ Promise.resolve(false), Promise.resolve(new Set()));
+        return { value: val };
     }
     else if (val instanceof Promise) {
-        // For a promise, we can just treat the same as an output that points to that resource. So
-        // we just create an Output around the Promise, and immediately apply the unwrap function on
-        // it to transform the value it points at.
-        const newOutput = new Output(
-            new Set(), val, /*isKnown*/ Promise.resolve(true), /*isSecret*/ Promise.resolve(false), Promise.resolve(new Set()));
-        return <any>(<any>newOutput).apply(output, /*runWithUnknowns*/ true);
+        return { value: createSimpleOutput(val.then(v => output(v))), isOutput: true };
```

Comment:
Why is this `isOutput: true` if it's a promise?

Reply:
Perhaps keeping whatever variant of the comment remains true would help.

Reply:
is it an invariant that `isOutput` is `true` if `value` is an Output?  If so... i don't really understand the benefit of this approach.  Can't we just do `Output.isINstance` on the results of outputRec?